{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "853cc0d7",
   "metadata": {},
   "source": [
    "# 03. logging 모듈로 로그를 출력하고 관리하기"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d89e5450",
   "metadata": {},
   "source": [
    "## logging 모듈 사용하기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "45807980",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] 2021-10-03 15:26:09,449 __main__ <ipython-input-2-abbad0da76d6>:28 크롤링을 시작했습니다.\n",
      "[INFO] 2021-10-03 15:26:09,449 __main__ <ipython-input-2-abbad0da76d6>:28 크롤링을 시작했습니다.\n",
      "[WARNING] 2021-10-03 15:26:09,450 __main__ <ipython-input-2-abbad0da76d6>:29 외보 사이트 링크는 크롤링하지 않습니다.\n",
      "[WARNING] 2021-10-03 15:26:09,450 __main__ <ipython-input-2-abbad0da76d6>:29 외보 사이트 링크는 크롤링하지 않습니다.\n",
      "[ERROR] 2021-10-03 15:26:09,451 __main__ <ipython-input-2-abbad0da76d6>:30 페이지를 찾을 수 없습니다.\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 1025, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 869, in format\n",
      "    return fmt.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 611, in format\n",
      "    s = self.formatMessage(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 580, in formatMessage\n",
      "    return self._style.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 422, in format\n",
      "    return self._fmt % record.__dict__\n",
      "TypeError: not enough arguments for format string\n",
      "Call stack:\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel_launcher.py\", line 16, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\traitlets\\config\\application.py\", line 845, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 619, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\base_events.py\", line 538, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\base_events.py\", line 1782, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\ioloop.py\", line 688, in <lambda>\n",
      "    lambda f: self._run_callback(functools.partial(callback, future))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\ioloop.py\", line 741, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 814, in inner\n",
      "    self.ctx_run(self.run)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 775, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 358, in process_one\n",
      "    yield gen.maybe_future(dispatch(*args))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 261, in dispatch_shell\n",
      "    yield gen.maybe_future(handler(stream, idents, msg))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 538, in execute_request\n",
      "    user_expressions, allow_stdin,\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 302, in do_execute\n",
      "    res = shell.run_cell(code, store_history=store_history, silent=silent)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 539, in run_cell\n",
      "    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2899, in run_cell\n",
      "    raw_cell, store_history, silent, shell_futures)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2944, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3170, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3361, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3441, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 38, in <module>\n",
      "    logging_example()\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 30, in logging_example\n",
      "    logger.error('페이지를 찾을 수 없습니다.') # error 레벨로 메시지 출력\n",
      "Message: '페이지를 찾을 수 없습니다.'\n",
      "Arguments: ()\n",
      "[ERROR] 2021-10-03 15:26:09,451 __main__ <ipython-input-2-abbad0da76d6>:30 페이지를 찾을 수 없습니다.\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 1025, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 869, in format\n",
      "    return fmt.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 611, in format\n",
      "    s = self.formatMessage(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 580, in formatMessage\n",
      "    return self._style.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 422, in format\n",
      "    return self._fmt % record.__dict__\n",
      "TypeError: not enough arguments for format string\n",
      "Call stack:\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel_launcher.py\", line 16, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\traitlets\\config\\application.py\", line 845, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 619, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\base_events.py\", line 538, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\base_events.py\", line 1782, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\ioloop.py\", line 688, in <lambda>\n",
      "    lambda f: self._run_callback(functools.partial(callback, future))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\ioloop.py\", line 741, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 814, in inner\n",
      "    self.ctx_run(self.run)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 775, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 358, in process_one\n",
      "    yield gen.maybe_future(dispatch(*args))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 261, in dispatch_shell\n",
      "    yield gen.maybe_future(handler(stream, idents, msg))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 538, in execute_request\n",
      "    user_expressions, allow_stdin,\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 302, in do_execute\n",
      "    res = shell.run_cell(code, store_history=store_history, silent=silent)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 539, in run_cell\n",
      "    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2899, in run_cell\n",
      "    raw_cell, store_history, silent, shell_futures)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2944, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3170, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3361, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3441, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 38, in <module>\n",
      "    logging_example()\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 30, in logging_example\n",
      "    logger.error('페이지를 찾을 수 없습니다.') # error 레벨로 메시지 출력\n",
      "Message: '페이지를 찾을 수 없습니다.'\n",
      "Arguments: ()\n",
      "[ERROR] 2021-10-03 15:26:09,459 __main__ <ipython-input-2-abbad0da76d6>:35 요청 중에 예외가 발생했습니다. : MissingSchema(\"Invalid URL '#invalid_url': No schema supplied. Perhaps you meant http://#invalid_url?\")\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 33, in logging_example\n",
      "    r = requests.get('#invalid_url', timeout=1)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\api.py\", line 75, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\api.py\", line 61, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\sessions.py\", line 528, in request\n",
      "    prep = self.prepare_request(req)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\sessions.py\", line 466, in prepare_request\n",
      "    hooks=merge_hooks(request.hooks, self.hooks),\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\models.py\", line 316, in prepare\n",
      "    self.prepare_url(url, params)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\models.py\", line 390, in prepare_url\n",
      "    raise MissingSchema(error)\n",
      "requests.exceptions.MissingSchema: Invalid URL '#invalid_url': No schema supplied. Perhaps you meant http://#invalid_url?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 33, in logging_example\n",
      "    r = requests.get('#invalid_url', timeout=1)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\api.py\", line 75, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\api.py\", line 61, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\sessions.py\", line 528, in request\n",
      "    prep = self.prepare_request(req)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\sessions.py\", line 466, in prepare_request\n",
      "    hooks=merge_hooks(request.hooks, self.hooks),\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\models.py\", line 316, in prepare\n",
      "    self.prepare_url(url, params)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\models.py\", line 390, in prepare_url\n",
      "    raise MissingSchema(error)\n",
      "requests.exceptions.MissingSchema: Invalid URL '#invalid_url': No schema supplied. Perhaps you meant http://#invalid_url?\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 1025, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 869, in format\n",
      "    return fmt.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 611, in format\n",
      "    s = self.formatMessage(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 580, in formatMessage\n",
      "    return self._style.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 422, in format\n",
      "    return self._fmt % record.__dict__\n",
      "TypeError: not enough arguments for format string\n",
      "Call stack:\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel_launcher.py\", line 16, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\traitlets\\config\\application.py\", line 845, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 619, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\base_events.py\", line 538, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\base_events.py\", line 1782, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\ioloop.py\", line 688, in <lambda>\n",
      "    lambda f: self._run_callback(functools.partial(callback, future))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\ioloop.py\", line 741, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 814, in inner\n",
      "    self.ctx_run(self.run)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 775, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 358, in process_one\n",
      "    yield gen.maybe_future(dispatch(*args))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 261, in dispatch_shell\n",
      "    yield gen.maybe_future(handler(stream, idents, msg))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 538, in execute_request\n",
      "    user_expressions, allow_stdin,\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 302, in do_execute\n",
      "    res = shell.run_cell(code, store_history=store_history, silent=silent)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 539, in run_cell\n",
      "    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2899, in run_cell\n",
      "    raw_cell, store_history, silent, shell_futures)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2944, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3170, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3361, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3441, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 38, in <module>\n",
      "    logging_example()\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 35, in logging_example\n",
      "    logger.exception('요청 중에 예외가 발생했습니다. : %r', e)\n",
      "Message: '요청 중에 예외가 발생했습니다. : %r'\n",
      "Arguments: (MissingSchema(\"Invalid URL '#invalid_url': No schema supplied. Perhaps you meant http://#invalid_url?\"),)\n",
      "[ERROR] 2021-10-03 15:26:09,459 __main__ <ipython-input-2-abbad0da76d6>:35 요청 중에 예외가 발생했습니다. : MissingSchema(\"Invalid URL '#invalid_url': No schema supplied. Perhaps you meant http://#invalid_url?\")\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 33, in logging_example\n",
      "    r = requests.get('#invalid_url', timeout=1)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\api.py\", line 75, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\api.py\", line 61, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\sessions.py\", line 528, in request\n",
      "    prep = self.prepare_request(req)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\sessions.py\", line 466, in prepare_request\n",
      "    hooks=merge_hooks(request.hooks, self.hooks),\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\models.py\", line 316, in prepare\n",
      "    self.prepare_url(url, params)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\models.py\", line 390, in prepare_url\n",
      "    raise MissingSchema(error)\n",
      "requests.exceptions.MissingSchema: Invalid URL '#invalid_url': No schema supplied. Perhaps you meant http://#invalid_url?\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 33, in logging_example\n",
      "    r = requests.get('#invalid_url', timeout=1)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\api.py\", line 75, in get\n",
      "    return request('get', url, params=params, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\api.py\", line 61, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\sessions.py\", line 528, in request\n",
      "    prep = self.prepare_request(req)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\sessions.py\", line 466, in prepare_request\n",
      "    hooks=merge_hooks(request.hooks, self.hooks),\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\models.py\", line 316, in prepare\n",
      "    self.prepare_url(url, params)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\requests\\models.py\", line 390, in prepare_url\n",
      "    raise MissingSchema(error)\n",
      "requests.exceptions.MissingSchema: Invalid URL '#invalid_url': No schema supplied. Perhaps you meant http://#invalid_url?\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 1025, in emit\n",
      "    msg = self.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 869, in format\n",
      "    return fmt.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 611, in format\n",
      "    s = self.formatMessage(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 580, in formatMessage\n",
      "    return self._style.format(record)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\logging\\__init__.py\", line 422, in format\n",
      "    return self._fmt % record.__dict__\n",
      "TypeError: not enough arguments for format string\n",
      "Call stack:\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\runpy.py\", line 193, in _run_module_as_main\n",
      "    \"__main__\", mod_spec)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\runpy.py\", line 85, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel_launcher.py\", line 16, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\traitlets\\config\\application.py\", line 845, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelapp.py\", line 619, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\platform\\asyncio.py\", line 199, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\base_events.py\", line 538, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\base_events.py\", line 1782, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\ioloop.py\", line 688, in <lambda>\n",
      "    lambda f: self._run_callback(functools.partial(callback, future))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\ioloop.py\", line 741, in _run_callback\n",
      "    ret = callback()\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 814, in inner\n",
      "    self.ctx_run(self.run)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 775, in run\n",
      "    yielded = self.gen.send(value)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 358, in process_one\n",
      "    yield gen.maybe_future(dispatch(*args))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 261, in dispatch_shell\n",
      "    yield gen.maybe_future(handler(stream, idents, msg))\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\kernelbase.py\", line 538, in execute_request\n",
      "    user_expressions, allow_stdin,\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\tornado\\gen.py\", line 234, in wrapper\n",
      "    yielded = ctx_run(next, result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 302, in do_execute\n",
      "    res = shell.run_cell(code, store_history=store_history, silent=silent)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\ipykernel\\zmqshell.py\", line 539, in run_cell\n",
      "    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2899, in run_cell\n",
      "    raw_cell, store_history, silent, shell_futures)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2944, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3170, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3361, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\master\\anaconda3\\envs\\multi\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3441, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 38, in <module>\n",
      "    logging_example()\n",
      "  File \"<ipython-input-2-abbad0da76d6>\", line 35, in logging_example\n",
      "    logger.exception('요청 중에 예외가 발생했습니다. : %r', e)\n",
      "Message: '요청 중에 예외가 발생했습니다. : %r'\n",
      "Arguments: (MissingSchema(\"Invalid URL '#invalid_url': No schema supplied. Perhaps you meant http://#invalid_url?\"),)\n"
     ]
    }
   ],
   "source": [
    "from logging import getLogger, Formatter, FileHandler, StreamHandler, DEBUG, ERROR\n",
    "import requests\n",
    "\n",
    "# 로거 :  __name__에는 실행 모듈 이름이 들어감(파일 이름)\n",
    "logger = getLogger(__name__)\n",
    "\n",
    "# 출력 형식\n",
    "default_format = '[%(levelname)s] %(asctime)s %(name)s %(filename)s:%(lineno)d %(message)s'\n",
    "default_formatter = Formatter(default_format)\n",
    "funcname_formatter = Formatter(default_format + ' (%funcName)s')\n",
    "\n",
    "# 로그 전용 핸들러 : 콘솔 출력 전용\n",
    "log_stream_handler = StreamHandler()\n",
    "log_stream_handler.setFormatter(default_formatter)\n",
    "log_stream_handler.setLevel(DEBUG)\n",
    "\n",
    "# 로그 전용 핸들러 : 파일 출력 전용\n",
    "log_file_handler = FileHandler(filename='crawler.log')\n",
    "log_file_handler.setFormatter(funcname_formatter)\n",
    "log_file_handler.setLevel(ERROR)\n",
    "\n",
    "# 로거에 핸들러와 레벨 설정하기\n",
    "logger.setLevel(DEBUG)\n",
    "logger.addHandler(log_stream_handler)\n",
    "logger.addHandler(log_file_handler)\n",
    "\n",
    "def logging_example():\n",
    "    logger.info('크롤링을 시작했습니다.') # info 레벨로 메시지 출력\n",
    "    logger.warning('외보 사이트 링크는 크롤링하지 않습니다.') # warning 레벨로 메시지 출력\n",
    "    logger.error('페이지를 찾을 수 없습니다.') # error 레벨로 메시지 출력\n",
    "    \n",
    "    try:\n",
    "        r = requests.get('#invalid_url', timeout=1)\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        logger.exception('요청 중에 예외가 발생했습니다. : %r', e)\n",
    "        \n",
    "if __name__ == '__main__':\n",
    "    logging_example()"
   ]
  },
  {
   "attachments": {
    "image-2.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeQAAABhCAYAAAAHpNImAAAgAElEQVR4Ae2d25ElSW5ER4A1owhUhxJQlJGGyvCDf6MHxRgalnvavL0dgcj7qMrqio80AP5AIDKjbnBmabt//OMf//j7Vc9ff/0Ve3X4q9a9U5/P3utnr3+nb3Fmed3f9nmX512eM/D+M/DHK19ydxl0+CvXvkuvz97rZ69/l+9w5nj/j8d5x+cdnzPw2jNwLuQX/huCOpznQnztAT1/8Od9njNwzsB3OQPnQj4XcvyPGb7LH8DZ5/mxP2fgnIG7nIF/Xsj8U11FHh0QXrHKwfF4RJ90cBrV73jVHa/alLuPGm3V5Bo7XDWerzys22ngq2fSTPzKB6c90uysS3RNqtHSmxqt1xMOf+L5oTxn4JyB73QGflzI/qOpteb6chz3Gm3hzl2pd/yslaKulXLF1N/hqvG88zi+qou7yjOH+wqf+qHRHqkPvMepf9erw73/qc+P8jkD5wx8hzPw40JOm9UfTM1L63WHdbj6Ndc5wImJU6zL1b+TVx/VdX0TnnwJ0zUSr5jmrJkw7YluB0u9EqY9Ne+0imvezaQ9T35+gM8ZOGfgu52BcyH/6z9DftWF4X3qQCVM8cQrpjkHNGHaE90OlnolTHtq3mkV17ybSXue/PwYnzNwzsB3OwPbF3K9GH5Uif6yruCq1Vx7ghMTp1iXq7/LywtH7Pqt8ORN2LSeejRn7YRpT3Q7WOqVMO2pead1nJqoPU5+fnzPGThn4LufgR8Xsv9Iel0vCozoL+8KrlrNtSc4MXGKdbn6u7y8cMSu3wpP3oTpeil3j9aVa63zJHwHU03lWmv/lCd98oMRU6+DnR/lcwbOGfiuZ+DHhVwvgB/W1Q/mI1zyOLaqnWPW3Y+m/i6nl/JgV2Lnd9zrWqMwcKKuPfH0UM9VjHWJV3oxX+ftZvE1Tn1+kM8ZOGfgO56Bny7knRew+rEtf/pRTp4Ow6+zdFrVrHL1dzl+5cGuxJW/OB7tmTyKaY4vYcUlfMImnjWv9FeP5mkt5U9+fojPGThn4Luega0LuX5EeX7HF8XeustCec9f9T6mvhP/7BzP9C/van3tvdId7vwQnzNwzsB3PgP/vJC/8ws4ez8/AOcMnDNwzsA5A3c4A+dCfvF/deYdPuqZ4fy4nDNwzsA5A1/vDPx0If/5559/6/NVPqjOXPlXmfvM+fX+YM43+5rfTH8XNOd7TpjymuMnFscDpnHlVV2Xf7a/m+vgr/m7+OVCTi+WA6ZRdYp7XjrHqna/1uRJRy80Ht3jfFfTl5h0cGkN5RJPvxWHZhVXfp1h1aPjVn7lVjN0vcGT13tT49mJeIjuASc6P9X4iK4HJzp/55qZNeq8inteOseqdr/W5ElHLzS7ER/RfbqW5ugmTHnN8Vd03OukwZ+0cBpXuuLSs+tX3clfc8FefY/bF7I31oOhueuqTrximqtfcc27nitc+3ruvVMf12itOb0dq5oHzZWI1/vSw3Gv0XXR9Vprjj9hcCmWnsf5q712/NpTc7wJg/OYtIppjjdhcHeLaVbFNE+zJ14xzdWvuOal8Vp9O7n7tdacXhOmvOYrf3Gu9XrywxPf5af/iZ9zEfPeP+1CrgE4XESGIoITwYkJTxj6LiaPYpprD3Bi4hSrPGlds6qTP2FX1pr8iU/Yam645EsYeo2dLuGKaU6/Xaz0k3biWfOucZo/8bqXjgcnqqdycGLHO75Te0+tNadXYdOjWnJi6lmc415Pfnjiu/z0P/GLXsj14TgcxO5jdjw40f3gROdTfUWb/GDaR3P4ih2+4lYe7d3lyZ+w1Qze+xF/5/HeXidfwtxX9a7Otbu+Xd2j/dOe7oB1+wYndrN2PDjR/eBE55+pvafWmrPGhCmv+cpfnGu9nvzwxHf56X/iN72Q9WBprgcCnKhcl1/Rdj0K1z6aq6fD3b/rUV2XpzUTtprBez/i7zze2+vkK8wf912tdR3Nr/bp9NpT805/Z7ybH5zY7SHximmuPcCJyj2be0+tNWedCVNec/wVHfc6afAnLZzGTtfh6q28dDzOnfpzL+N6/w//K2s+LpGPrJEPrBg5HH6tyUu74tFpxKPY1dx7eE2/q3j5kqew7mEtYueH14i26628+sjhq/YeaK5G7Yl3F0Of4mo+59J6qWeHuf/V/bt134X7flgHPO0PrrSJpwe81uT0III/GnUOenQYvEbV+kxaa67+yrWHc/BXcNd2a3f4rt91p/6cy/llF/LqA6bDopjm2gecCFc1DxjRteC7MfkTVv0SnjBde+JVm/LkT1g3327Pyd+tmfortuvb1WlvzdWvOZqEwa1i8u1iq76fyaX5ax5wYjdj4hXTXHuAE+Gq5gGbovdwvfKau+6z6t2ZOl2H+352de479cdc0J92IdcH5nAQ/aODEzte8U6rmi7vvLt4p9P1djSq9zz5E1a+Dt/pOfl3e++u9ajOfdQ6n+aJB5ti6lOehCds6v9ZfDcrOLGbr+PBie4HJ3a846nueqBVXnP4ioXvPOrRvOu7o9nxMqP2I3/WT58TP+bi7d7zuZD/9d/UtTrQHae45t3LLnxX1/VI/oRdWesRf+fp5gbf9SVdwujrUbWao9vFVvoVl/qjv1vsZgUndnN3PDjR/eDEjnc81V0PtMprDn8ldv4O196dpsPxFs8DpnHyo93VoT/xYy/ohy9k/bCapw/Y8eBE94ITO17xTqsazydPx4MTvW+qr2h3/V3PDve+nQ6cqL6EKd/lyfcMVutM/oln1qTr+uPp+K6X+u6Sp1kV0zzN3PHgRPeCEzve8VR3PdAqrzk8sbjpQetx8sG7jxo+RdWQayyP1prTr7CVTj0n/9iLmPf9aReyHgzNGayi4prDOQauPXby1Md9rtFac/d5fUXr3qo7v+Nep16KuV5rzfEkDG4Vk28X6/pO/onv+oInP1zFxCdMPXfK06yKaZ7mTrximqtfcc1LU7Vj6k05Ho/0w9P17XB8U9zx72hW66z8xaVH+638pZt47XXy11/a2xeyf2j9GM5pzUdWrHL1P6LBk/o4NtU+G7X7wH1+xTV3f9XuTZoVtvJPa6/6Mhs9XAtOdH63Ln/S0peYNCsMH9G14ETnVzUej+pZcaq7Y+6zV61zJh6sdOQa1f+IBo/3ebSu2fBqDlaxw1Wzynf8O5pn13jG/+x8q7UPN1/gv1zI9UF4vsoLZF7iV5n7zDkf0POOzjt6xRmo3wb6aA5G5Ddkiug9PurzPl29mr3zKP6sX3ud/PV/mz9dyOcFv/4Fn3d63uk5A+cMnDNwzsDOGTgX8vnfQ/7xTw47B+Zozg/LOQPnDJwz8J4z8NOF/Ndff/2tz11eus5U+V3mOnO851Ce93re69UzoL8LmtNnwpTXHD+xOB4wjSuv6rr8s/3dXAf/mL/JXy7k6cWnA5Ow6gNesXt0PdUorjk9FXtFrmt3a6jG11Su85dnxXnPVK/8OkPyTtjKr9xqhp01XOO9qV23qvEQXQtOdH6q8RFdD050/qvUNb/PmrDSgLPnFLWX8orv5Oqt3D2KaY5uwpTXHH9Fx71OGvxJC6dxpSsuPbt+1Z38Yy7Yq+/5ZRdyOkhgxNVwrvEab4fDPxJTT8dWtXM1g2NV8zw648qf1ruyzsrvXNrftBazv6KXrzX1nHjv5/Xkn3jvd+e620uH114S53t0jdeun2r3a605fSZMec1X/uJc6/Xkhye+y0//E+95EfNdbnEhXzmEnZYNPRJTT8U01/7gxMQpVnnSumZVJ3/Crqw1+ROfsNXccMmXMPQaO13CFdOcfrtY6SftxLPmV4jdXjq8ez+61+Td8WkPz72n1prjK2x6VEtOTD2Lc9zryQ9PfJef/id+kwu5PrQfJmpidxg6PuEJ6/o+g+s6mmvPDi9Nx3W49l3lyZ+w1Qze/xF/5/HeXidfwtxX9a7Otbu+Xd2j/dOe7oaldwBGZGZqIrjHju9w96favVprjnfClNd85S/OtV5Pfnjiu/z0P/FcyL8cWj8UVw5hp/Wez9a6jubat8NL03Edrn1XefInbDWD93/E33m8t9fJV5g/7rta6zqaX+3T6bWn5p3+q+BpL2BE9kJNBPfY8R3u/lS7V2vN8U6Y8prjr+i410mDP2nhNHa6Dldv5aXjce7U976M6/u87F9Z87H14JBzQDwmD1hF/BOm/CtyX9dr1riKly95Cuse1iJ2fniNaLveyquPHL5q74HmatSeeHcx9Cmu5nMurZd6dpj7X92/W/cjcN9bralYytP+k87nV41zqdZ14DsMXqNqfW2tNVd/5drDOfgruGu7tTt81++6U9/zcv6wC3l1ALrDlvCErXpf5VL/hFXfhCdMZ5h41aY8+RPWzbfbc/J3a6b+iu36dnXaW3P1a44mYXCrmHy72KrvXbhpL8qTE7s9dHyHpz6TVnnNU6/PwHZn6nQd7nvZ1bnv1Pe4oF9+IdeH5VB47D46OucTnjD3PVp3vXfxTqfz7GhU73nyJ6x8Hb7Tc/Lv9t5d61Gd+6h1Ps0TDzbF1Kc8CU/Y1P8OfJrbMWqP3fzonO9w11U9aZXXXHsVvvOoR/Ou745mx1t9Ol2H69orv+tOfY8L2L/DuZD/9d/UtTrwHae45v6Std7VqUfz5E9YeTpc+610K/+K8/5a7/qSLmHaW3PVao5mF1vpV1zqj/7OMc3tGLXHbl/onO9w11U9aZXXPPWasM7f4dqv03Q43uJ5wDROfrS7OvQn3utifsuFXB9ZD9d0SDo+4Ql79lBNPTsenLgzxxVt6pf8CeMbpB6OTf7EJ8z7pjr5nsFqjck/8cyZdF1/PB3f9VLfHfM0d4eBE7v9dHyHpz6TVnnNvVdx0+Me6skHj94jfIpoiyPX2OGloR+5+k5+rwt3+h63uJDTQeoOYIdPG13xOz1do7Xmq3XSPie9891ajnvtfbx2vdaa40sY3Com3y7W9Z38E9/1BU9+uIqJT5h67pqnuTsMnLjak2u8XnnhypOe4rWf5nhdo/hu3vVV/45G9Z6v/MWlR3us/KWbeO118o+/zN92IevH3z0Eeti6w7Dbq/MnXNfV3LUdp7jm7q+6+ITvYiv/tPa0xsqv3GqGnTWS5tn+k3/i00xg7qWGrwhGVO4r5TW/z5sw9qzRfV7zbrp+rr9Sa0/NtUeHq2aV7/h3NM+u8Yz/2flWax/u+Qv8lwu5PhjPXV4w8xDvMteZ4/kDeN7heYevOAP120AfzcGI/IZMEb3HR33ep6tXs3cexZ/1a6+Tf/zf5k8X8vkAH/8Bzjs/7/ycgXMGzhk4Z6DOwLmQz/8e8o9/sjg/CudH4ZyBcwbOGfi8M/DThfzf//XH3/p8lw+je678u+z77PPz/vDOuz/v/iudAf1d1Jw9TJjymuMnFscDpnHlVV2Xf7a/mwv8lwsZQqNuosvRK69Y4frAVUyeDk/ahCV/t/7OLKr5HXN9N6v3ic7fATjReequN/wUV37WXmlW/Vd+5R7tX2snr/emXs3qHB7iVd71XtOXeJV3/Z3q2hPzdHniFeO9EOEqas8JT9qEpb6s3el1bc/Vm/yKaU6fCVNec/wVHfc6afAnLZzGla649Oz6Vfdo/mEXsg+oL4aXsNLAqU+xDlcNecWkX+Hq/d3y9C4cW9XOpfdYGp5H3h/etFa33pV1vK/WmtMzYXAplp7H+au9dvzaU3O8CYPzmLSKaY43YXB3izprlzOz8rtYeR7xaf/J77zX9NqN7tdac/pNmPKar/zFudbryQ9PfJef/s/GT7uQa3BeTkVy3dAVbKVNnK4/ran875in96OY5rp/cGLiFKs8aV2zqpM/YVfWmvyJT9hqbrjkSxh6jZ0u4YppTr9drPSTduJZ865R5+9yZld+hRWHtiI5HuV3sJU/cV1/XWuVe0+tNadHYdOjWnJi6lmc415Pfnjiu/z0fza+9EJms0SG89pxeKLz1BVdo5hz1ETt0+VXtF2P3wHX96C57q3DS9NxHa59V3nyJ2w1g/d/xN95vLfXyZcw91W9q3Ptrm9X92j/tKc7YLrvLtc9q0Zx3ws6j+jAqbte6IjoqYngr4jeU2vNWWvClNd85S/OtV5Pfnjiu/z0fzaeC9n+v6y7D/bsi/5qfn0Pmus+Orw0Hdfh2neVJ3/CVjN4/0f8ncd7e518hfnjvqu1rqP51T6dXntq3unvjOv8XV7zwxHZk9eOwxOdp67oGsWcoyZqn2dz76m15qwzYcprjr+i414nDf6khdPY6TpcvZWXjse5V9SXLmSG9sggE46uItrdHK/6EqY8ORH9Kl7Rrvp8Zc7fgdfs7SpevuQprHtYi9j54TWi7Xorrz5y+Kq9B5qrUXvi3cXQp7iaz7m0XurZYe5/df9u3Xfh7KeLrOu849QV0e7meNWXMOXJiegfjdWHhx7Uuobm6Cqq1jVaa65+7+Ec/BXctd3aHb7rd92j9csuZN+Q1pX7owO7Fk7xXUw95ETtUZjjxScM33eIaf8J695Vp+XdTTy6LiZ/wrr5Ut9H/J0n9Vds17er096aq19zNAmDW8Xk28VWfT+TY/4u1mxwzKl15f6gc6/7VOdauM4DTlR9YY7Dpzhpldc89foMbHemTtfhvpddnft26w+7kH0g3ZjmpaMmqncHQ+NR++g6iuNR7Lvk3d538U6n729Ho3rPkz9h5evwnZ6Tf7f37lqP6txHrfNpnniwKaY+5Ul4wqb+n8UzaxfTHtEmzjHVKue4cpXzuI7aI3oiPPUqTlrlNdeehe886tG867uj2fFWn07X4br2yu+6R+utCzkN4huo2h+Gcq3jzlMT0ac5EobPo/ZJvg5z3+9Y867S3jpOcc1TD7BdHXqPyZ+w8nX4Ts/Jv9t7d60d3ZU1Vas56+xiK/2KS/3R3zH6vKkuTB/24VrHnacmoq+4g6HxqH26Xq6hphe1R+U1d91O3fk7XHt2mg7HWzwPmMbJj3ZXh/5qfMmF3A0JTvThwInKF9bhqqu804ETd32u+93r7v2w744HJ6JfxSva1Cf5E1beDve+nQ6cqL6EKd/lyfcMVutM/oln1qTr+uPp+K6X+u6U+7xaa64zgxOVqxycqHxhHa467aO4+lOfzqc9NO96oFFec3hicdOD1uPkg3cfNXyKqiHXWB6tNadfYSudeh7Nv8WFnF6kvmR9ee9+4brWXfKdPbtGa82nPV3Rpl6d33GvUy/FXK+15ngSBreKybeLdX0n/8R3fcGTH65i4hOmnrvlPq/Wmuvc4ETlKgcnKl9Yh6tO+yjufu/lvHq7HI9Hn8HXol+Hw09xx7+jWa2z8heXHu238pdu4rVXym97IXebSxtOWPKXjifxHVb47/zwTjz6npVXTnHNVUNePPkjceWf1p7WW/mVW82ws0bSPNt/8k98mgnMvdTwFcGIyn2FvObWObXWPGme5VPPCSve162aJ/Ha82qua2mufTpcNat8x7+jeXaNZ/zPzvfLhVwNeVaD/U4c+yX+Tns7e/m9/4+p833P9/2IM1C/jayjORiR39Apovf4qM/7dPVq9s6j+LN+7ZXyny7kJDjY+YM/Z+CcgXMGzhk4Z+D9Z+BcyL/5v44+f0Tv/yM67/i843MGzhl4xRn46UL+n3/797/1ecUCH9FDZ678I9Y8a5w/wHMGvs4Z0N8FzfmGE6a85viJxfGAaVx5Vdfln+3v5jr4a/4WfrmQ04vlgKVY+g6n18QnHdhOf9dqvZv7jMmnGueVq9x56hWHZhVXfp1h1aPjVn7lVjN0vcGT13tT49mJeIjuASc6P9X4iK4HJzp/55qZU6y5O5w9TXzSge30V22X+wyuKx5M811MPZrjr+i410mDP2nhNK50xaVn16+6k7/mgr36Hrcv5FXjdEgU05w+jq1q56pHwlY466aYejm2qp1Lc5SGJ80wYXjTWt16U0/lva/WmuNJGFyKpedx/mqvHb/21BxvwuA8Jq1imuNNGNzd4jRr4hXTnL05tqqdqx4Jo/dOdL/WmtNrwpTXfOUvzrVeT3544rv89D/xcy5i3vvbLuRagMNDZFEiOBGcCE4EJyY8Yei7mDyKaa49wImJU6zypHXNqk7+hF1Za/InPmGrueGSL2HoNXa6hCumOf12sdJP2olnzbvGNL/O2vHgRPVUDk58lHffTu1raq05vQqbHtWSE1PP4hz3evLDE9/lp/+J50L+5dByKDh8RHBiwhOG/krUPpprjw4vTcd1uPZd5cmfsNUM3v8Rf+fx3l4nX8LcV/WuzrW7vl3do/3Tnu6ATfvueHCi7wWc+Cjvvp3a19Rac3pNmPKar/zFudbryQ9PfJef/ieeC/mXQ8uh4PARwYkJTxj6K1H7aK49Orw0Hdfh2neVJ3/CVjN4/0f8ncd7e518hfnjvqu1rqP51T6dXntq3unvjE/zdzw40fcITnyUd99O7WtqrTm9Jkx5zfFXdNzrpMGftHAaO12Hq7fy0vE4d+rPvYzr/W//K2s+okY+YGHkRMU03+VLh4+IVznFVrjrVrWv5zXeq3g3X/XpHtYipjUTpmt1vfERWYOouPdAczVqT7y7GPoUV/M5l9ZLPTvM/a/u3637LjzNr3vUnBkU03yXLx0+Il7lFJvy6sODllrX0BxdRdW6RmvN1e89nIO/gru2W7vDd/2uO/XnXM7bF/LqA9Vh8Ef1zvnh8Rov+ORHXxGPYlfy5E9Yt1anZYaJR9fF5E9YN1/q+4i/86T+iu36dnXaW3P1a44mYXCrmHy72KrvZ3Jpfp2neH+u8qonZ13vDY5uJ04e5TXf6f0Rmt2ZOl2H++y7Oved+mMu6JddyP7B9MNrjk4xzeErghMTp5h6HN+p0zqrnq73Oq25o0k+sORP2GpuehEf8XceenZx17er21kn9UpY1wu88yQ8YfS5W5xmTbximrM3xTSHrwhOTJxiqzz1UL3ymrumuOlRj+Zd3x3Njrf6dLoO17VXfted+mMuYH/Pb7uQ9eN3hwWc6MOBEzte8U6rmpSvfB2nuOapP9iuDr3H5E9Y+Tp8p+fk3+29u9aO7sqaqtWcdXaxlX7Fpf7o7xanWTsenOj7Aic+yrsv1d0aaJXXHP5K7Pwdrr07TYfjLZ4HTOPkR7urQ3/ix17M50LeuLi6QwxO3Dm8V7SpX/InrLwd7n07HThRfQlTvsuT7xms1pn8E8+sSdf1x9PxXS/13SWfZu14cKLvB5z4KO++VHdroFVec3hicdOD1uPkg3cfNXyKqiHXWB6tNadfYSudek7+sRcx7/sWF3I6KHpwNGfw5Okw9aS8669a12ituXpSfkV7xe99vU69FHO91prjSRjcKibfLtb1nfwT3/UFT364iolPmHrulE+zdjw40fekuOb+zpyjT4fDeyx9et61XlrfMa+v7umKP+3d1/P6Sn/Xnvr1l/b2hbz62N1HBif6B3S8ah7Vug4u4QlD30XW9Oh65ZVTXHPVkBdP/khc+ae1p/VWfuVWM+yskTTP9p/8E59mAnMvNXxFMKJyd8+ZOUX2lvZQ+h0er/YHu+JXz9WcWd+5nq7Rzbej6byr2Vce5ab1J157nfwDLuT6IDxf5YUzL/GrzH3mfP2BPu/0vNN0Buq3AVxzMCK/IVNE7/FRn/fp6tXsnUfxZ/3a6+Sv/1v76Z+Qzwt+/Qs+7/S803MGzhk4Z+CcgZ0zcC7k87+H/OOfHHYOzNGcH5ZzBs4ZOGfgPWfgpwv5f//jP//W57z0/3/p+k4qP+/lPYfxvNfzXs8Z+JpnQH8XNed7TpjymuMnFscDpnHlVV2Xf7b/lwu5G7RwXoQPrTh518e93hc/0fskPxo8kwY9caVHQ7yixfMVor67bo+q8T0p1/nLs+K8Z6pXfp0heSds5VduNcPOGq7x3tSuW9V4iK4FJzo/1fiIrgcnOv8VamavqPMqTq685u4tDk+K6kXrGLX6wTx267vuaq1rT2tMPGurrsvRVlRNqjtshRenj6/hXPH+uEbrq/n2heyDaq05AySsuA6ffBPvfb1e+TstHo1XtOq7c5725Niqdq726ljVPI+8C7zel16Oe42ui67XWnP8CYNLsfQ8zl/ttePXnprjTRicx6RVTHO8CYO7Y/R5tdac2RNWXIdPvon3vl6v/J0WzyPRe2qtOb0nTHnNV/7iXOv15IcnvstP/yluXcjTkBOvQ3RaNI/wnSfhuxjzeEx+13y1Ou1JMc11b+DExClWedK6ZlUnf8KurDX5E5+w1dxwyZcw9Bo7XcIV05x+u1jpJ+3Es+ZdY5pf9z3xuq9Oi+YRvvMkfBdjnkejr6O15vQvbHpUS05MPYtz3OvJD098l5/+U/xWFzIvm8jL8Ro8xSva5P8qmO5Tc52/w0vTcR2ufVd58idsNYP3f8Tfeby318mXMPdVvatz7a5vV/do/7SnO2DdvsGJPmvCE6a+R/jO4zg1kXW9Bn8mek+tNWeNCVNe85W/ONd6Pfnhie/y03+KT13INL+yiU479VrxU0/3ut5r9Cle0Sb/V8F0n5rr/B1emo7rcO27ypM/YasZvP8j/s7jvb1OvsL8cd/VWtfR/GqfTq89Ne/0d8an+Ts+4QnTvT/CTx76oyM6Tv2KuFrDuVpvwpTXXGd13OtunRWu/Ve6tJZ78Zd2V+893nYhdwN1OIM9wk8e7+16r9GneEWb/F8B8z16zR6u4uVLnsK6h7WInR9eI9qut/LqI4ev2nuguRq1J95dDH2Kq/mcS+ulnh3m/lf379Z9F+778XUSn7DydTg9H+Enj/d2vdfor8bqw4OXWtfQHF1F1bpGa83V7z2cg7+Cu7Zbu8N3/a7r6pddyDWwPt2C08Ye4SdPzeIarTXv5ga/osXzlWLaX8JqTwlPmO5/4lWb8uRPWDffbs/J362Z+iu269vVaW/N1a85moTBrWLy7WKrvp/Jpfl1nuL9UV7znV6q9zz5Ezb51KO5+3brqYfymu/2f7dud6ZO1+E+967Off7sMwYAAAMRSURBVNQvu5BpSOwG6/DJt+KnnuV1jdaas04Xr2i7HnfFu73t4p1O97ujUb3nyZ+w8nX4Ts/Jv9t7d61Hde6j1vk0TzzYFFOf8iQ8YVP/z+KnWROfsO5d6L46H5rEJww90TVaa47+apx6KK+5rlP4zqMezbu+O5odb/XpdB2ua6/8ruvqpy5khiT6IglPmPoe4TuP4pX7w7qqA+viFW3X4474al8dp7jmq/3t6roeyZ+w8ne49+50HX6l9+5aO7rVPCt/8u1i9E36FbfS47tL7GYFJ/q8CU+Y+h7hO4/ilfvDuqoDuxqnHsprfnWd0nf+Dtc1Ok2H4y2eB0zj5Ee7q0Pv8VtcyN1LAif6y0n1FW3y3xGb9tTx4MSdvV3Rpn7Jn7Dydrj37XTgRPUlTPkuT75nsFpn8k88syZd1x9Px3e91HeXvJsVnOjzJjxh6nuE7zzgRF2ncnCi81fqqYfymvsaxU2Pe6gnHzx6j/Apoi2OXGOHl4Z+5Oq7mm9dyGkhHVBzHSDhCZs8O7z31Vrz1KvjVUt+RYvn7nFnT67RWvNpr1e0qVfnd9zr1Esx12utOZ6Ewa1i8u1iXd/JP/FdX/Dkh6uY+ISp5265z6u15jp3whM2eXZ476u15qlXx6t2J68+6SmvrqG59u1w1azyHf+O5tE1qnd6tN+0/sRvX8i1qA6zM0RaPGE7vdCs/I/Opz7NWVPjan3VfaVc96y576HjFNfc/VUXn/BdbOWf1p7WWPmVW82ws0bSPNt/8k98mgnMvdTwFcGIyn2VnNkr6sxewyU8YegrPsOXl2enJ2vh8ag9ns1Za7VH1Tyy3o5/R7Na+93+qf8vF3IZeFaDfyeO90H8Tns/e/35v+v2vI/zPs4Z+PUM1G8j70VzMCK/oVNE7/FRn/fp6tXsnUfxZ/0/Xcja+OS/HrrzTs47OWfgnIFzBs4ZeNcZOBey/a99vOtFn77nj/icgXMGzhk4Z2B1Bs6FfC7kH/+qaXVQDnd+SM4ZOGfgnIH3noFzIZ8L+VzI5wycM3DOwDkDNzgD/wdM4+V6+H68YAAAAABJRU5ErkJggg=="
    }
   },
   "cell_type": "markdown",
   "id": "d383c073",
   "metadata": {},
   "source": [
    "## 딕셔너리 형식으로 설정 작성하기\n",
    "- settings.py\n",
    "\n",
    "```python\n",
    "\"\"\"설정 파일\"\"\"\n",
    "import os\n",
    "\n",
    "BASE_DIR = os.path.realpath(os.path.dirname(__file__))\n",
    "LOG_DIR = os.path.join(BASE_DIR, 'logs')  # 로그 파일 디렉터리\n",
    "\n",
    "# 로그 파일 디렉터리가 없다면 생성하기\n",
    "if not os.path.exists(LOG_DIR):\n",
    "    os.mkdir(LOG_DIR)\n",
    "LOGGING_CONF = {\n",
    "    'version': 1,  # 필수\n",
    "    # logger 설정 처리가 중복되면 덮어씌우기\n",
    "    'disable_existing_loggers': True,\n",
    "    # 출력 형식 설정\n",
    "    'formatters': {\n",
    "        'default': {  # 디폴트 형식\n",
    "            '()': 'colorlog.ColoredFormatter',  # colorlog 라이브러리 적용\n",
    "            'format': '\\t'.join([\n",
    "                \"%(log_color)s[%(levelname)s]\",  # 로그 레벨\n",
    "                \"asctime:%(asctime)s\",  # 로그 출력 날짜\n",
    "                \"process:%(process)d\",  # 로그 출력을 실행한 프로세스 이름\n",
    "                \"thread:%(thread)d\",  # 로그 출력을 실행한 프로세스 ID\n",
    "                \"module:%(module)s\",  # 로그 출력을 실행한 프로세스 모듈 이름\n",
    "                \"%(pathname)s:%(lineno)d\",  # 로그 출력을 실행한 모듈의 경로와 줄 번호\n",
    "                \"message:%(message)s\",  # 로그 출력할 메시지\n",
    "            ]),\n",
    "            'datefmt': '%Y-%m-%d %H:%M:%S',  # asctime으로 출력할 로그 출력 날짜 형식\n",
    "            # 로그 레벨에 따라 색 적용하기\n",
    "            'log_colors': {\n",
    "                'DEBUG': 'bold_black',\n",
    "                'INFO': 'white',\n",
    "                'WARNING': 'yellow',\n",
    "                'ERROR': 'red',\n",
    "                'CRITICAL': 'bold_red',\n",
    "            },\n",
    "        },\n",
    "        'simple': {  # 로그를 적게 출력하는 간단한 형태의 형식\n",
    "            '()': 'colorlog.ColoredFormatter',  # pip install colorlog\n",
    "            'format': '\\t'.join([\n",
    "                \"%(log_color)s[%(levelname)s]\",\n",
    "                \"%(asctime)s\",\n",
    "                \"%(message)s\",  # 로그 레벨, 날짜, 메시지만 출력\n",
    "            ]),\n",
    "            'datefmt': '%Y-%m-%d %H:%M:%S',\n",
    "            'log_colors': {\n",
    "                'DEBUG': 'bold_black',\n",
    "                'INFO': 'white',\n",
    "                'WARNING': 'yellow',\n",
    "                'ERROR': 'red',\n",
    "                'CRITICAL': 'bold_red',\n",
    "            },\n",
    "        },\n",
    "        'query': {  # SQL 쿼리 출력 전용 형식\n",
    "            '()': 'colorlog.ColoredFormatter',\n",
    "            'format': '%(cyan)s[SQL] %(message)s',  # 쿼리만 출력\n",
    "        },\n",
    "    },\n",
    "    # 로그 출력 대상에 따른 핸들러 설정\n",
    "    'handlers': {\n",
    "        'file': {  # 파일에 로그를 출력할 핸들러 설정\n",
    "            'level': 'DEBUG',  # logger.level이 DEBUG 이상일 때만 출력\n",
    "            # 로그 크기가 일정량을 넘으면 새로운 로그 파일을 생성하는 핸들러\n",
    "            'class': 'logging.handlers.RotatingFileHandler',\n",
    "            # 로그 파일 경로 지정\n",
    "            'filename': os.path.join(LOG_DIR, 'crawler.log'),\n",
    "            'formatter': 'default',  # 디폴트 형식으로 로그 출력\n",
    "            'backupCount': 3,  # 오래된 로그 파일은 3개만 남김\n",
    "            'maxBytes': 1024 * 1024 * 2,  # 로그 크기가 2MB를 넘을 경우 새로운 로그 파일 생성\n",
    "        },\n",
    "        'console': {  # 터미널에 로그를 출력하는 핸들러의 로그 형식 지정\n",
    "            'level': 'DEBUG',\n",
    "            'class': 'logging.StreamHandler',  # 터미널에 로그를 출력할 핸들러\n",
    "            'formatter': 'default',  # 디폴트 형식으로 로그 출력\n",
    "        },\n",
    "        'console_simple': {  # 터미널에 로그를 출력하는 핸들러의 간단한 로그 형식 지정\n",
    "            'level': 'DEBUG',\n",
    "            'class': 'logging.StreamHandler',\n",
    "            'formatter': 'simple',  # 간단한 형식 지정\n",
    "        },\n",
    "        'query': {  # 터미널에 SQL 쿼리 로그를 출력할 핸들러\n",
    "            'level': 'DEBUG',\n",
    "            'class': 'logging.StreamHandler',\n",
    "            'formatter': 'query',  # SQL 쿼리 출력 전용 형식\n",
    "        },\n",
    "    },\n",
    "    'root': {  # 디폴트 설정\n",
    "        'handlers': ['file', 'console_simple'],  # 위에 설정한 file, console 설정으로 출력\n",
    "        'level': 'DEBUG',\n",
    "    },\n",
    "    # 로그 이름, 핸들러, 로그 레벨 설정\n",
    "    'loggers': {\n",
    "        # logging.getLogger(__name__)의 __name__으로 참조되는 이름이 키로 사용됨\n",
    "        'celery': {\n",
    "            'handlers': ['console', 'file'],\n",
    "            'level': 'WARNING',  # Celery 로그는 WARNING 이상만 출력\n",
    "            'propagate': False,  # 로그 이벤트를 루트 로거에 전달하지 않게 지정\n",
    "        },\n",
    "        'my_project': {  # my_project.py 모듈에서 사용할 때의 로거\n",
    "            'handlers': ['console', 'file'],\n",
    "            'level': 'DEBUG',\n",
    "            'propagate': False,\n",
    "        },\n",
    "    },\n",
    "}\n",
    "```\n",
    "\n",
    "- my_logging.py\n",
    "\n",
    "```python\n",
    "\"\"\"로그 전용 모듈\"\"\"\n",
    "import logging.config\n",
    "\n",
    "import settings\n",
    "\n",
    "def get_my_logger(name):\n",
    "    logging.config.dictConfig(settings.LOGGING_CONF)\n",
    "    return logging.getLogger(name)\n",
    "\n",
    "logger = get_my_logger(__name__)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    \"\"\"my_logging를 사용해봅니다.\"\"\"\n",
    "    logger.debug('DEBUG 레벨입니다.')\n",
    "    logger.info('INFO 레벨입니다.')\n",
    "    logger.warning('WARNING 레벨입니다.')\n",
    "    logger.error('ERROR 레벨입니다.')\n",
    "    logger.critical('CRITICAL 레벨입니다.')\n",
    "```\n",
    "\n",
    "- my_logging.py 실행 결과\n",
    "![image-2.png](attachment:image-2.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45f7475b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python Multi",
   "language": "python",
   "name": "multi"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
